### PR TITLE
Weights & Biases (wandb) integration

### PR DESCRIPTION
**Pull request by alejandromarcu**
This PR allows to track the training in wandb, as well as save and load model parameters.

**When Training**
- If you don't want to use wandb, either pass to the training `--no-wandb` or run from the command line `wand disabled` and then it won't be used until re-enabled
- To use wandb you can:
-- set in your environment WANDB_API_KEY to your API key
-- when running the training script, it will give instructions on how to create an account or log in
-- run 'wandb login` form the command line

A new wandb run will be created, containing all the metadata to reproduce and showing graphs in real time of the loss, rewards and epsilon and more.
E.g. [Run data image](https://github.com/user-attachments/assets/dc23fa8c-1b96-498a-8561-64c255799ae8), [Run metrics image](https://github.com/user-attachments/assets/f5708519-7442-454a-a4e4-d2590f781f07)

When the run finishes, the model file will be uploaded to wandb as an artifact under a name such as `dexp_B5W2_mv1`.  The first part is the model name, then board size and wall max, and the "model version", which needs to be changed every time we make changes that are not backwards compatible (e.g. you change the number of parametrs).
If the file generated is the same than an existing one, wandb won't create a duplicate, but you may have different versions if you change the hyperparameters, e.g.:

image

**When Playing** 
- I removed the pretrained subclasses, so now the player name is for example dexp, not dexppretrained.  When the model is instantiated with `training_mode=False`, then the parameters are going to be loaded
- If the player is just the friendly name of the agent, it will try to load locally, e.g. `--players dex greedy`.
- The agent can contain an "alias" for wandb, in the above image they are v2, latest, v1 or v0.  To make it use a model from wandb, just pass an alias, e.g. `--players dex:latest dex:v1` (yeah, you can make it play against a different model of itself!).  One issue that I just realized is that if there's more than 1 person training the same model, when they play they may not grab their own model when using latest.  Maybe we should use an alias with the nickname (notice that no more than one thing can have the same alias, so when you put it in a new one, it's removed from the older one)
- One thing that I suggest is to have a process for aliases, e.g. we can create one that is `default` that we'll use to indicate a good working model, or we may have `best` to track which one is producing the best results.

### CODE REVIEW

File: deep_quoridor/src/train.py [hunk 1]
```diff
@@ -42,31 +49,80 @@ def end_game(self, game, result):
...
+                )
+        self.episode_count += 1
+
+    def save(self, agent: AbstractTrainableAgent, filename: str):
+        artifact = wandb.Artifact(f"{agent.model_id()}", type="model", metadata={"episodes": self.total_episodes})
+        artifact.add_file(local_path=filename)
+        artifact.save()
+
+
 class SaveModelEveryNEpisodesPlugin(ArenaPlugin):
```
Comment (muralx) [inline deep_quoridor/src/train.py:92]: why merge the two. Keep one for wandb and one for files. Cleaner and we can use the one we want.
- Note we should liklely move these plugin out of the file, we are having too many of them in the same file.

**Review by muralx — APPROVED** [review]
Thanks Cucu, I left some comments. Address the ones you consider are OK to do now. We can do the others later (or not)

File: deep_quoridor/src/agents/core/agent.py [hunk 2]
```diff
@@ -36,6 +36,10 @@ class AgentRegistry:
 
     @staticmethod
     def create(friendly_name: str, **kwargs) -> Agent:
+        if ":" in friendly_name:
```
[the remark this answers is not in the document]
Comment (alejandromarcu) [reply deep_quoridor/src/agents/core/agent.py:39]: @muralx After your PR, for dex the way to specify the player is for example: `dexppretrained-101`.
You're suggesting to migrate that too, right? so something like `dexppretrained:flags=101` (or if you prefer I could make it explicit, e.g. `dexppretrained:use_rotate_board=true`

File: deep_quoridor/src/agents/core/trainable_agent.py [hunk 3]
```diff
@@ -232,13 +256,39 @@ def load_model(self, path):
...
+            if os.path.exists(filename):
+                print(f"Loading pre-trained model from {filename}")
+                self.load_model(filename)
+            else:
+                raise FileNotFoundError(
+                    f"Model file {filename} not found. Please run training or provide an alias to load from wandb, e.g. 'dex:latest'"
+                )
         else:
-            raise FileNotFoundError(f"Model file {model_path} not found. Please provide the weights file.")
+            api = wandb.Api()
```
[the remark this answers is not in the document]
Comment (alejandromarcu) [reply deep_quoridor/src/agents/core/trainable_agent.py:274]: You mean `self.load_model(filename)`?

File: deep_quoridor/src/agents/core/agent.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alejandromarcu) [reply deep_quoridor/src/agents/core/agent.py:39]: actually, using `,` is not problematic with the shell, so we could use that as the separator between arguments, and use `:` after the name, e.g. `dexp:a=1,b=2`

[the remark this answers is not in the document]
Comment (muralx) [reply deep_quoridor/src/agents/core/agent.py:39]: Yes, I meant you can just use your approach and we can get rid of the change I did. Latest suggestion looks good, I also thought using ","  would be a problem. You can keep the numbers for now, I might get rid of them depending on what I can do with wandb

File: deep_quoridor/src/agents/core/trainable_agent.py [hunk 4]
```diff
@@ -54,11 +59,30 @@ def __init__(
         self.optimizer = self._create_optimizer()
         self.criterion = self._create_criterion()
         self.replay_buffer = ReplayBuffer(capacity=10000)
-        self.training_mode = False
+        self.training_mode = training_mode
         self.episodes_rewards = []
         self.train_call_losses = []
         self.reset_episode_related_info()
 
+        if not training_mode:
```
[the remark this answers is not in the document]
Comment (muralx) [reply deep_quoridor/src/agents/core/trainable_agent.py:67]: Note also that we might need to load from wandb one model and then write back a different model.

For instance use model1 as based, and train and get model b.

File: deep_quoridor/src/agents/core/trainable_agent.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (muralx) [reply deep_quoridor/src/agents/core/trainable_agent.py:274]: Yeah, something like that, so actual loading messaging and handling is in one place.

**Review by muralx — APPROVED** [review]

File: deep_quoridor/src/agents/core/agent.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alejandromarcu) [reply deep_quoridor/src/agents/core/agent.py:39]: I implemented this here: https://github.com/jonbinney/deep_rabbit_hole/pull/146/files
